### PR TITLE
Custom comparator

### PR DESCRIPTION
**Pull request by georeith**
Expose the `compareWith` function that allows you to set a custom function to use for `is()` comparison when comparing non Immutable objects as opposed to `equals()` or setting it to `null` to force strict object equality.

This solves this: https://github.com/facebook/immutable-js/issues/814

And any other situation where you don't want `equals` to be used on your objects. It does not change the behaviour of Immutable collection comparison nor does it cause any breaking changes to existing code. As the tests show.

### CODE REVIEW

**Comment by lacker** [comment]
`compareWith` seems like a problematic thing to expose because it works indirectly, it's like it just operates with side effects. At least if I understand it properly. Also maybe this just timed out at this point. I am going to close this pull request but let me know if you think there is still something we should do here.

**Comment by georeith** [comment]
@lacker It is not pure to speak but then again neither is the default of using `.equals()` if it exists on the item. Which is the whole reason this needs to exist because sometimes you are using items that you don't control the `.equals()` function for. You can't override it because X library expects a certain behaviour from it.

This is still not pure with or without `compareWith` its just `compareWith` lets you work around the side effects if you need to. I am happy for another solution but nothing else is being suggested.

In short this is just a fix around the already un pure behaviour which has causes side effects.

**Comment by elliotbonneville** [comment]
I'm with @georeith on this one. `compareWith()`'s current behavior is causing loads of problems for me as well, in terms of working with other libraries which also expose `#equals()` functions.

**Comment by lacker** [comment]
It seems like the right way to do this would be to create a collection with a custom comparator - that's how most languages handle this sort of data structure. Then it's per-collection rather than using `compareWith` impurely.
